src/agent/src/storage/mod.rs

```diff
@@ -55,6 +54,86 @@ pub struct StorageContext<'a> {
     sandbox: &'a Arc<Mutex<Sandbox>>,
 }
 
+/// An implementation of generic storage device.
+pub struct StorageDeviceGeneric {
+    path: String,
+    initialized: bool,
+}
+
+impl std::fmt::Debug for StorageDeviceGeneric {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("StorageDeviceGeneric")
+            .field("path", &self.path)
+            .finish()
+    }
+}
+
+impl Default for StorageDeviceGeneric {
+    fn default() -> Self {
+        StorageDeviceGeneric {
+            path: "".to_string(),
+            initialized: false,
+        }
+    }
+}
+
+impl StorageDeviceGeneric {
+    /// Create a new instance of `StorageStateCommon`.
+    pub fn new(path: String) -> Self {
+        StorageDeviceGeneric {
+            path,
+            initialized: true,
+        }
+    }
+}
+
+impl StorageDevice for StorageDeviceGeneric {
+    fn path(&self) -> &str {
+        &self.path
+    }
+
+    fn cleanup(&self) -> Result<()> {
+        if !self.initialized {
+            return Ok(());
+        } else if self.path.is_empty() {
+            return Err(anyhow!("mountpoint path is empty"));
+        }
+
+        let path = Path::new(&self.path);
+        if !path.exists() {
+            return Ok(());
+        }
+
+        if matches!(is_mounted(&self.path), Ok(true)) {
+            let mounts = vec![self.path.clone()];
+            remove_mounts(&mounts)?;
+        }
+        if matches!(is_mounted(&self.path), Ok(true)) {
+            return Err(anyhow!("failed to umount mountpoint {}", self.path));
+        }
+
+        if path.is_dir() {
+            let is_empty = path.read_dir()?.next().is_none();
+            if !is_empty {
+                return Err(anyhow!("directory is not empty when clean up storage"));
+            }
```

Comment:
Maybe the 116-119 is unnecessary because the `remove_dir` can ensure that the directory is empty.

Reply:
There may be multiple filesystems mounted at the same destination directory, and umount() only unmounts one level, so hope to give a clear error message for such a case.